mpc.c

```diff
@@ -2120,6 +2120,12 @@ mpc_parser_t *mpc_tok_braces(mpc_parser_t *a, mpc_dtor_t ad)   { return mpc_tok_
 mpc_parser_t *mpc_tok_brackets(mpc_parser_t *a, mpc_dtor_t ad) { return mpc_tok_between(a, ad, "{", "}"); }
 mpc_parser_t *mpc_tok_squares(mpc_parser_t *a, mpc_dtor_t ad)  { return mpc_tok_between(a, ad, "[", "]"); }
 
+mpc_parser_t *mpc_sepby1(mpc_fold_t f, mpc_parser_t *sep, mpc_parser_t *a) {
+  return mpc_and(2, f,
+    a, mpc_many(f, mpc_and(2, mpcf_snd_free, sep, mpc_copy(a), free)),
+    free);
+}

```

Comment:
@orangeduck I've realized that the fold function is applied twice. This makes it hard (or maybe impossible) to build an array of structures with a custom fold function.

Do you have any advice on how to fix this?

Seems like we need a new repeat parser, but I'm lost on how to add those.

Reply:
I guess you mean that the fold function when used in `mpc_many` will get a list of entries, while when used in `mpc_and` it will just get the first item and the rest of the elements?

Yeah I see the problem.

The easiest thing would be to let people give two different fold functions to use in the `mpc_and` and the `mpc_many`.  

There also might be a way to implement this in terms of `mpc_many1` - in particular if you allow a trailing separator...

Otherwise I am not totally sure, and yes perhaps the only way is to implement a custom repeat parser type.

Reply:
@orangeduck Many thanks for the reply. I'm more familiar with the repeat parsers code now. I'm working on adding a new one for sepby1.

Preliminary work for that on https://github.com/orangeduck/mpc/pull/166.

Reply:
Thanks a lot for working on this addition 👍